optics_functions/coupling.py

```diff
@@ -255,44 +271,111 @@ def closest_tune_approach(df: TfsDataFrame, qx: float = None, qy: float = None,
     deltaq = qx_frac - qy_frac  # fractional tune split
-    exponential_term = ((deltaq * df[S] / (df.headers[LENGTH] / PI2)) - (df[f"{PHASE_ADV}{X}"] - df[f"{PHASE_ADV}{Y}"]))
-    return 4 * deltaq * df['F1001'] * np.exp(1j * exponential_term)
+    location_term = np.exp(1j * PI2 * (deltaq * df[S] / (df.headers[LENGTH] / PI2)))
+    return _cta_persson_alt(df, qx_frac, qy_frac) * location_term
+
+
+def _cta_hoydalsvik(df: TfsDataFrame, qx_frac: float, qy_frac: float) -> Series:
+    """ Closest tune approach calculated by Eq. (14) in
+    [HoydalsvikEvaluationOfTheClosestTuneApproach2021]_ .
+    This is like the persson estimate but divided by 1 + 4|F1001|^2 ."""
+    return _cta_persson(df, qx_frac, qy_frac) / (1 + 4 * df[F1001].abs() ** 2)
 
 
-def _cta_calaga(df: TfsDataFrame, qx_frac: float, qy_frac: float) -> Series:
+def _cta_hoydalsvik_alt(df: TfsDataFrame, qx_frac: float, qy_frac: float) -> Series:
+    """ Closest tune approach calculated by Eq. (14) without the s-term in
+    [HoydalsvikEvaluationOfTheClosestTuneApproach2021]_ .
+    This is like the persson_alt estimate but divided by 1 + 4|F1001|^2 ."""
+    return _cta_persson_alt(df, qx_frac, qy_frac) / (1 + 4 * df[F1001].abs() ** 2)
+
+
+def _cta_teapot(df: TfsDataFrame, qx_frac: float, qy_frac: float) -> Series:
     """Closest tune approach calculated by Eq. (27) in [CalagaBetatronCoupling2005]_ .
     If F1010 is not given, it is assumed to be zero.
     """
-    f_diff = df["F1001"].abs() ** 2
-    with suppress(KeyError):
-        f_diff -= df["1010"].abs() ** 2
+    f_diff = df[F1001].abs() ** 2
+    with suppress(KeyError):  # this is the only estimate that uses sum resonance
+        f_diff -= df[F1010].abs() ** 2
 
     return (
         (np.cos(PI2 * qx_frac) - np.cos(PI2 * qy_frac))
         / (np.pi * (np.sin(PI2 * qx_frac) + np.sin(PI2 * qy_frac)))
         * (4 * np.sqrt(f_diff) / (1 + 4 * f_diff))
     )
+
+
+def _cta_teapot_franchi(df: TfsDataFrame, qx_frac: float, qy_frac: float) -> Series:
+    """Closest tune approach calculated by Eq. (12) in
+    [HoydalsvikEvaluationOfTheClosestTuneApproach2021]_ .
+    This is the teapot approach with the Franchi approximation. """
+    return 4 * (qx_frac - qy_frac) * df[F1001].abs() / (1 + 4 * df[F1001].abs() ** 2)
+
+
+def _get_weights_from_lengths(df: TfsDataFrame) -> Tuple[float, np.array]:
+    """Coefficients for the `persson` method. """
+    # approximate length of each element (ds in integral)
+    s_periodic = np.zeros(len(df) + 1)
+    s_periodic[1:] = df[S].to_numpy()
+    s_periodic[0] = df[S][-1] - df.headers[LENGTH]
+
+    # weight ds/(2*pi*R) * N (as we take the mean afterwards)
+    weights = np.diff(s_periodic) / df.headers[LENGTH] * len(df)
+    return weights
+
+
+def check_resonance_relation(df: DataFrame, to_nan: bool = False) -> DataFrame:
+    """Checks that F1001 >= F1010. If desired, sets the invalid points to NaN.
```

Comment:
Should be absolute values here I believe

Reply:
Could also expand that the "invalid" are where the condition is not met, and explain why we set them to NaN

Reply:
I feel the invalid is obvious and the NaN is because they are invalid. I don't see what to explain